diff --git a/src/resmom/mom_main.c b/src/resmom/mom_main.c
index 73c9d8a..b3de418 100644
--- a/src/resmom/mom_main.c
+++ b/src/resmom/mom_main.c
@@ -7138,6 +7138,12 @@ int main(
     }
 #endif  /* NVIDIA_GPUS */
 
+#ifdef PENABLE_LINUX_CGROUPS
+  snprintf(log_buffer, sizeof(log_buffer),
+    "cgroup support enabled");
+  log_ext(-1, __func__, log_buffer, LOG_DEBUG);
+#endif
+
   main_loop();
 
   if (mom_run_state == MOM_RUN_STATE_KILLALL)
diff --git a/src/server/pbsd_main.c b/src/server/pbsd_main.c
index 530abd7..8ba1bad 100644
--- a/src/server/pbsd_main.c
+++ b/src/server/pbsd_main.c
@@ -1931,6 +1931,12 @@ int main(
 
 #endif
 
+#ifdef PENABLE_LINUX_CGROUPS
+  snprintf(log_buf, sizeof(log_buf),
+    "cgroup support enabled");
+  log_ext(-1, __func__, log_buf, LOG_DEBUG);
+#endif
+
   /*==========*/
   main_loop();
   /*==========*/
